mmv1/templates/terraform/custom_delete/storage_folder_delete.go.tmpl

```diff
@@ -0,0 +1,88 @@
+        log.Printf("Error! %s : %s\n\n", bucket, deleteErr)
+        return deleteErr
+    }
+    // GCS requires that a folder be empty (have no objects or object
+    // versions) before it can be deleted.
+    log.Printf("[DEBUG] GCS Folder attempting to forceDestroy\n\n")
+
+    // Create a workerpool for parallel deletion of resources. In the
+    // future, it would be great to expose Terraform's global parallelism
+    // flag here, but that's currently reserved for core use. Testing
+    // shows that NumCPUs-1 is the most performant on average networks.
+    //
+    // The challenge with making this user-configurable is that the
+    // configuration would reside in the Terraform configuration file,
+    // decreasing its portability. Ideally we'd want this to connect to
+    // Terraform's top-level -parallelism flag, but that's not plumbed nor
+    // is it scheduled to be plumbed to individual providers.
+    wp := workerpool.New(runtime.NumCPU() - 1)
+
+    for _, object := range res.Items {
+        log.Printf("[DEBUG] Found %s", object.Name)
+        object := object
+
+        wp.Submit(func() {
+            log.Printf("[TRACE] Attempting to delete %s", object.Name)
+            if err := config.NewStorageClient(userAgent).Objects.Delete(bucket, object.Name).Generation(object.Generation).Do(); err != nil {
+                deleteObjectError = err
+                log.Printf("[ERR] Failed to delete storage object %s: %s", object.Name, err)
+            } else {
+                log.Printf("[TRACE] Successfully deleted %s", object.Name)
+            }
+        })
+    }
+
+    // Wait for everything to finish.
+    wp.StopWait()
+}
+
+if listError != nil {
+    return fmt.Errorf("could not delete non-empty folder due to error when listing contents: %v", listError)
+}
+log.Printf("[DEBUG] force_destroy value: %#v", d.Get("force_destroy").(bool))
+foldersList, err := config.NewStorageClient(userAgent).Folders.List(bucket).Prefix(name).Do()
+if err != nil {
+    return err
+}
+if len(foldersList.Items) == 1 || d.Get("force_destroy").(bool) {
+    log.Printf("[DEBUG] folder names to delete: %#v", name)
+    items := foldersList.Items
+    for i := len(items) - 1; i >= 0; i-- {
+        err = transport_tpg.Retry(transport_tpg.RetryOptions{
+            RetryFunc: func() error {
+                err = config.NewStorageClient(userAgent).Folders.Delete(bucket, items[i].Name).Do()
+                return err
+            },
+            Timeout:              d.Timeout(schema.TimeoutDelete),
+            ErrorRetryPredicates: []transport_tpg.RetryErrorPredicateFunc{transport_tpg.Is429RetryableQuotaError},
+        })
+        if err != nil {
+            return err
+        }
+    }
+
+    log.Printf("[DEBUG] Finished deleting Folder %q: %#v", d.Id(), name)
+} else {
+    deleteErr := fmt.Errorf("Error trying to delete folder without force_destroy set to true")
```

Comment:
Nit: "... X subfolders still exist within %name, use force_destroy = true to delete all subfolders", or similar, to provide more info?

Reply:
yes even if subfolders exist and force_destroy set to true we can delete the sub folders  else don't delete

Reply:
updated message